File: plugins/agent-box-harness/runtime/access-entry.mjs
/**
 * AgentBox Harness access entry — the plugin's one production entry for connecting to a Harness.
 *
 * It owns the five responsibilities the access layer has and no others:
 *
 *   1. **discovery** — `harnesses` lists the brands this plugin can connect to, and a connect
 *      reports the launch context discovery derived for that brand next to the one actually used,
 *      so a mismatch is visible instead of silently redirected;
 *   2. **connect** — `connect` starts the Harness with the launch context the caller names and
 *      transmits **no ACP frame** while doing it; one establishment at a time, reserved before the
 *      launch is awaited, because "one connection per process" has to be enforced rather than hoped
 *      for;
 *   3. **transport** — every other line on stdin is an ACP frame and goes to the Harness as
 *      written; every line the Harness writes comes back as written, including requests it raises
 *      itself and errors it answers with. Which lines are "other" is decided by shape, not by one
 *      field: a frame that declares itself JSON-RPC is relayed even if it carries an `op` of its own;
 *   4. **state** — `status` reports the connection and process, and the end of the transport is
 *      reported as its own event when the process exits;
 *   5. **close** — `close` releases the process this entry started and nothing else, and says so only
 *      once the OS has stopped answering for that pid. A Harness that ignores the graceful signal is
 *      escalated; a release that still cannot be confirmed is reported as a failure with the process
 *      identity kept, never as `released: true`.
 *
 * What is deliberately absent: no session, turn, snapshot or model-catalog management; no
 * `permissionRoundTrip`, `permissionTimeoutMs` or brand `permissionMode`, because those are how the
 * old chain answered the Harness on the caller's behalf; no product model-id aliasing on the way
 * out, because that rewrites an ACP parameter. The old envelope and the native-driver branch that
 * served it are retired — see `REMOVALS.md` for what went and where it is recoverable from.
 *
 * **One connection per process.** A byte-transparent relay cannot multiplex two Harnesses over one
 * pair of stdin/stdout lines without stamping each frame with a discriminator, and a stamp is
 * exactly the envelope this entry exists to stop using. A caller that needs two connections starts
 * two entries; Workcore already owns one channel run per execution.
 *
 * Security posture carried over from the retired entry: the Worker must say it is isolated
 * (`AGENTBOX_SIDECAR_ISOLATED=1`) or the caller must ask for native mode explicitly, never both;
 * the reused snapshot is hash-verified before anything runs; adapter environment names are validated
 * and refused when they look like credentials; and a credential value read from the Worker's secret
 * path is redacted out of every diagnostic this process emits.
 */
import { createHash } from "node:crypto"
import { readFileSync } from "node:fs"
import { spawn } from "node:child_process"
import path from "node:path"
import { fileURLToPath } from "node:url"
import readline from "node:readline"
import { harnessLaunchContext, isKnownHarness, listHarnesses } from "../harnesses/index.mjs"
import { openAcpConnection, PROCESS_GROUP_OWNERSHIP } from "./access-transport.mjs"

const here = path.dirname(fileURLToPath(import.meta.url))
const snapshotRoot = path.resolve(here, "..", "third_party", "harness_remote")
const CREDENTIAL_PATH = "/runtime/secret/credential"
const ENVIRONMENT_KEY = /^[A-Z][A-Z0-9_]{0,63}$/
const SENSITIVE_ENVIRONMENT_KEY = /(TOKEN|SECRET|KEY|PASSWORD|CREDENTIAL|AUTH)/i

/** The whole control vocabulary. Anything else on stdin is an ACP frame, not a request. */
const CONTROL_FIELDS = new Set(["op", "id"])
const CONNECT_FIELDS = new Set(["harness", "launch", "directory", "credentialEnvironment"])
const LAUNCH_FIELDS = new Set(["command", "args", "environment"])
/**
 * The two fields that mark a message as belonging to the JSON-RPC conversation rather than to this
 * entry's own control envelope. `jsonrpc` is what a conforming peer declares; `method` is what a
 * request carries. A reply is covered too, because a message with a `result` or `error` is an answer
 * to a frame the caller sent, and an entry that swallowed it would strand the caller's own turn.
 */
const JSON_RPC_MARKERS = ["jsonrpc", "method", "result", "error"]

/**
 * Control or data? Decided by the shape of the message, never by one field of it.
 *
 * A top-level string `op` on its own is not enough: `op` is a legal extension member of an ACP frame,
 * and an entry that treated such a frame as its own would hijack traffic it was asked to relay — the
 * caller would see its frame vanish and a `UNKNOWN_OP` back instead of the Harness's answer. So a
 * message that declares itself part of the JSON-RPC conversation goes to the Harness unchanged, and
 * only a bare control object (this entry's own envelope, which carries no `jsonrpc` and no `method`)
 * is handled here. The rule is one-directional on purpose: it can only ever move a message from
 * control to the data channel, never the other way.
 */
function isControlRequest(message) {
  return typeof message.op === "string"
    && !JSON_RPC_MARKERS.some((field) => message[field] !== undefined)
}
/**
 * Names the retired envelope used for one job or another. They are refused by name rather than
 * ignored: a caller that still sends `permissionTimeoutMs` believes it configured a timeout that
 * decides, and the honest answer to that belief is that this entry has no such knob.
 */
const RETIRED_CONNECT_FIELDS = [
  "profile", "stateDirectory", "permissionRoundTrip", "permissionTimeoutMs", "preferredAuthMethod",
  "driver", "title", "sessionId", "text", "model", "attachments", "requestId", "decision", "scope",
]

function verifyProvenance() {
  const source = JSON.parse(readFileSync(path.join(snapshotRoot, "SOURCE.json"), "utf8"))
  for (const file of source.files) {
    const expected = file.patched_sha256 ?? file.current_sha256 ?? file.upstream_sha256
    const actual = createHash("sha256").update(readFileSync(path.join(snapshotRoot, file.path))).digest("hex")
    if (actual !== expected) throw new Error(`PROVENANCE_MISMATCH: ${file.path}`)
  }
  return source
}

class Refused extends Error {
  constructor(code, detail) {
    super(detail ? `${code}: ${detail}` : code)
    this.code = code
  }
}

async function main() {
  const argv = process.argv.slice(2)
  const native = argv.length === 1 && argv[0] === "--native"
  if (argv.length > (native ? 1 : 0)) {
    fail({ code: "SIDECAR_MODE_INVALID" })
    return
  }
  if (native && process.env.AGENTBOX_SIDECAR_ISOLATED === "1") {
    fail({ code: "SIDECAR_MODE_CONFLICT" })
    return
  }
  if (!native && process.env.AGENTBOX_SIDECAR_ISOLATED !== "1") {
    fail({ code: "SIDECAR_ISOLATION_REQUIRED" })
    return
  }
  let source
  try {
    source = verifyProvenance()
  } catch (error) {
    fail({ code: "PROVENANCE_UNREADABLE", message: String(error?.message ?? error) })
    return
  }

  let connection = null
  let credentialValue = null
  /**
   * Every Harness process this entry started and has not confirmed gone.
   *
   * `connection` is only the handle the transport is currently routed to, so it cannot be the
   * bookkeeping for releases: a handle produced by a connect that then failed, or one whose release
   * was refused by the process, would otherwise belong to nobody while its process kept running.
   */
  const owned = new Set()
  /**
   * The establishment in flight, or `null`. Checking `connection` and then awaiting the launch is not
   * enough on its own: `connection` is assigned after the launch, so a second connect arriving during
   * that window would pass the check and start a second Harness whose handle the first then
   * overwrites. One reservation, taken before the first await, is what makes "one connection per
   * process" true rather than merely intended.
   */
  let opening = null

  function reserveOpening() {
    let settle
    const settled = new Promise((resolve) => { settle = resolve })
    // `done` clears the reservation before it wakes anyone waiting on it, so a waiter re-reading
    // `opening` sees either nothing or a genuinely later establishment.
    const reservation = { settled, done: () => { if (opening === reservation) opening = null; settle() } }
    opening = reservation
    return reservation
  }

  /** Wait for the establishment in flight to reach its outcome; a no-op when there is none. */
  async function settleOpening() {
    while (opening) await opening.settled
  }

  /**
   * Release every Harness process tree still owned, and report it.
   *
   * `released` is true only when every process in every one of those trees was confirmed gone by the
   * OS — `ESRCH` for each pid the release could account for (see `releaseProcessTree`); a survivor or
   * a probe the OS answered any other way keeps its entry in `owned` and its identity and per-process
   * receipt in the report, so the caller is never handed `released: true` without evidence.
   */
  async function releaseOwned() {
    const serving = connection
    const reports = []
    for (const handle of [...owned]) {
      const report = await handle.close()
      if (report.released) owned.delete(handle)
      if (handle === serving && report.released) {
        connection = null
        credentialValue = null
      }
      reports.push(report)
    }
    const unreleased = reports.filter((report) => !report.released)
    const primary = reports.find((report) => report.connectionId === serving?.connectionId) ?? reports[0]
    return {
      connectionId: primary?.connectionId ?? null,
      released: unreleased.length === 0,
      processId: primary?.processId ?? null,
      signalUsed: primary?.signalUsed ?? null,
      // The reclaiming is what round 5 asks a report to prove, so it travels with the verdict rather
      // than staying inside the handle that wrote it.
      tree: primary?.tree ?? null,
      ...(unreleased.length
        ? {
            unreleased: unreleased.map((r) => ({
              connectionId: r.connectionId, processId: r.processId, tree: r.tree ?? null,
            })),
          }
        : {}),
    }
  }

  function safeText(value, maximum) {
    let text = String(value ?? "")
    if (credentialValue) text = text.replaceAll(credentialValue, "[REDACTED]")
    return text.slice(0, maximum)
  }
  function emit(message) {
    process.stdout.write(`${JSON.stringify(message)}\n`)
  }
  function fail({ id = null, code, message }) {
    emit({ id, ok: false, error: { code, message: message ?? code } })
  }
  function refusedError(error, id) {
    emit({
      id: id ?? null,
      ok: false,
      error: { code: error?.code ?? "SIDECAR_OP_FAILED", message: safeText(error?.message ?? error, 500) },
    })
  }

  async function connect(request) {
    if (connection) throw new Refused("ACP_ALREADY_CONNECTED", String(connection.connectionId))
    if (opening) throw new Refused("ACP_CONNECT_IN_PROGRESS")
    const unknown = Object.keys(request).filter((key) =>
      !CONTROL_FIELDS.has(key) && !CONNECT_FIELDS.has(key) && !key.startsWith("_"))
    if (unknown.length) {
      const retired = RETIRED_CONNECT_FIELDS.filter((name) => unknown.includes(name))
      throw new Refused(retired.length ? "ACP_CONNECT_FIELD_RETIRED" : "ACP_CONNECT_FIELD_UNKNOWN", retired.join(", "))
    }
    const harness = request.harness
    if (typeof harness !== "string" || !harness) throw new Refused("HARNESS_REQUIRED")
    if (!isKnownHarness(harness)) throw new Refused("HARNESS_UNDISCOVERED", harness)
    const launch = request.launch
    if (launch === null || typeof launch !== "object" || Array.isArray(launch)) {
      throw new Refused("ADAPTER_LAUNCH_REQUIRED")
    }
    const extraLaunchFields = Object.keys(launch).filter((key) => !LAUNCH_FIELDS.has(key) && !key.startsWith("_"))
    if (extraLaunchFields.length) throw new Refused("ADAPTER_LAUNCH_FIELD_UNKNOWN", extraLaunchFields.join(", "))
    if (typeof launch.command !== "string" || !launch.command) throw new Refused("ADAPTER_LAUNCH_REQUIRED")
    if (launch.args !== undefined && (!Array.isArray(launch.args)
        || launch.args.some((arg) => typeof arg !== "string"))) {
      throw new Refused("ADAPTER_ARGS_INVALID")
    }
    const environment = launch.environment ?? {}
    if (environment === null || typeof environment !== "object" || Array.isArray(environment)) {
      throw new Refused("ADAPTER_ENVIRONMENT_INVALID")
    }
    for (const [key, value] of Object.entries(environment)) {
      if (!ENVIRONMENT_KEY.test(key) || SENSITIVE_ENVIRONMENT_KEY.test(key)
          || typeof value !== "string" || value.length > 8192 || /[\0]/.test(value)
          || /^sk-[A-Za-z0-9_-]+$/.test(value)) {
        throw new Refused("ADAPTER_ENVIRONMENT_INVALID", key)
      }
    }
    const credentialEnvironment = request.credentialEnvironment
    if (credentialEnvironment != null) {
      if (typeof credentialEnvironment !== "string" || !ENVIRONMENT_KEY.test(credentialEnvironment)) {
        throw new Refused("CREDENTIAL_ENVIRONMENT_INVALID")
      }
      credentialValue = readFileSync(CREDENTIAL_PATH, "utf8").trim()
      if (!credentialValue || credentialValue.length > 4096 || /[\r\n\0]/.test(credentialValue)) {
        credentialValue = null
        throw new Refused("CREDENTIAL_MATERIAL_INVALID")
      }
    }
    // The launch is put in a process group of its own (`detached` on POSIX, which makes the child a
    // session and group leader) so that "what this connection started" is a fact the OS can answer
    // rather than a list this process has to keep: a bridge that exits ahead of the Agent it started
    // leaves that Agent running, and `kill(-rootPid, …)` is the only thing that reaches it.
    //
    // On Windows the reused client keeps its own default `spawn` — it special-cases `.cmd` entries
    // there, and replacing it would quietly turn that handling off — so no group is claimed and the
    // release reports the single process it can account for instead. The two halves read one
    // constant, so the claim and the launch cannot drift apart.
    const groupLeadership = PROCESS_GROUP_OWNERSHIP ? { detached: true } : {}
    const needsEnvironment = Object.keys(environment).length > 0 || Boolean(credentialValue)
    let spawnProcess
    if (process.platform !== "win32" || needsEnvironment) {
      spawnProcess = (command, args, options = {}) => spawn(command, args, {
        ...options,
        ...groupLeadership,
        ...(needsEnvironment ? {
          env: {
            ...process.env, ...environment,
            ...(credentialValue ? { [credentialEnvironment]: credentialValue } : {}),
          },
        } : {}),
      })
    }

    // The reservation is taken here, in the same synchronous step as the launch: from the moment this
    // await yields, another connect can be handled, and it must find the establishment already
    // claimed rather than an empty `connection`.
    const reservation = reserveOpening()
    let opened
    try {
      opened = await openAcpConnection({
        harness,
        launch: { command: launch.command, args: launch.args ?? [] },
        directory: request.directory ?? process.cwd(),
        spawnProcess,
        redact: safeText,
      })
      // Claimed before anything else can fail or be assigned: from here there is a real Harness
      // process whose release this entry is responsible for, whether or not the setup below succeeds.
      owned.add(opened)
      connection = opened
      opened.onFrame((_message, line) => {
        if (line === undefined) emit(_message)
        else process.stdout.write(`${line}\n`)
      })
      opened.onStderr((line) => emit({ event: "stderr", data: { line }, connectionId: opened.connectionId }))
      opened.onMalformed((message) => {
        emit({ event: "transport_malformed", data: { message }, connectionId: opened.connectionId })
      })
      opened.onEnd((end) => {
        // Reported as it happened: no process is claimed to be alive after its exit, and
        // no `cancelled`/`refused` outcome is invented for requests still in flight.
        emit({ event: "transport_end", data: end, connectionId: opened.connectionId, harness })
      })
      return {
        connectionId: opened.connectionId,
        harness,
        transport: opened.transport,
        // What was actually launched, next to what discovery derived for this brand. A deployment that
        // pins an offline artifact entry wants to see `source: "npx"` in `discovered` and `explicit`
        // here — a redirect would hide the difference.
        launch: { ...opened.launch, source: "explicit" },
        discovered: harnessLaunchContext(harness),
        state: opened.status(),
        provenance: { commit: source.commit, ref: source.ref },
      }
    } finally {
      reservation.done()
    }
  }

  async function handle(request) {
    const { op, id } = request
    if (op === "harnesses") {
      return {
        provenance: { commit: source.commit, ref: source.ref },
        harnesses: listHarnesses(),
        transport: connection?.transport ?? null,
      }
    }
    if (op === "connect") return connect(request)
    if (op === "status") {
      if (!connection) {
        if (opening) throw new Refused("ACP_CONNECT_IN_PROGRESS")
        throw new Refused("ACP_NOT_CONNECTED")
      }
      return connection.status()
    }
    if (op === "close") {
      // A close that arrives while an establishment is in flight waits for it and releases what it
      // produced. The alternative — refusing because there is no connection yet — would be answerable
      // only while a freshly spawned Harness stayed owned by nobody, and this entry's last
      // responsibility is that it never leaves one running.
      await settleOpening()
      if (!connection && !owned.size) throw new Refused("ACP_NOT_CONNECTED")
      return releaseOwned()
    }
    throw new Refused("UNKNOWN_OP", op)
  }

  const rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity })
  for await (const line of rl) {
    const trimmed = line.trim()
    if (!trimmed) continue
    let message
    try {
      message = JSON.parse(trimmed)
    } catch {
      fail({ code: "ENVELOPE_MALFORMED" })
      continue
    }
    if (message === null || typeof message !== "object" || Array.isArray(message)) {
      fail({ code: "ENVELOPE_MALFORMED" })
      continue
    }
    if (isControlRequest(message)) {
      // Control and transport share this stream without a wrapper, so the one ordering rule is that
      // a control op never waits for the transport: `close` must still be answerable after the
      // Harness has died, and a caller may ask for `status` while a turn is open.
      void Promise.resolve().then(() => handle(message)).then(
        (result) => emit({ id: message.id ?? null, ok: true, result }),
        (error) => refusedError(error, message.id),
      )
      continue
    }
    // Not a control op: an ACP frame for the Harness, forwarded as written. No id is minted, no
    // field is added, and a frame nobody answers stays unanswered — the entry has no branch that
    // writes a result or a refusal on the Harness's behalf.
    if (!connection) {
      emit({ id: message.id ?? null, ok: false, error: { code: "ACP_NOT_CONNECTED", message: "no connection; send op:\"connect\" first" } })
      continue
    }
    try {
      connection.writeLine(trimmed)
    } catch (error) {
      emit({
        id: message.id ?? null,
        ok: false,
        error: { code: "ACP_CHANNEL_DEAD", message: safeText(error?.message ?? error, 500) },
      })
    }
  }
  // The host hung up. Release every process this entry owns before leaving, so an exiting caller
  // cannot strand a Harness — which includes an establishment that is still in flight when the host
  // stops talking, because its process appears a moment after the decision to exit is made. A release
  // the OS will not confirm is reported with its pid rather than called `released`: this process has
  // no stronger signal left than `SIGKILL`, and leaving is the only move that does not hang the
  // caller forever.
  try {
    await settleOpening()
    if (connection || owned.size) {
      const report = await releaseOwned()
      if (!report.released) emit({ event: "process_release_unconfirmed", data: report })
    }
  } catch (error) {
    emit({ event: "process_release_unconfirmed", data: { message: safeText(error?.message ?? error, 500) } })
  }
  process.exit(0)
}

function fail({ code, message }) {
  process.stdout.write(`${JSON.stringify({ ok: false, error: { code, message: message ?? code } })}\n`)
}

main().catch((error) => {
  fail({ code: "SIDECAR_FATAL", message: String(error?.stack ?? error).slice(0, 2000) })
  process.exit(1)
})

File: apps/server/src/ordessa_server/acp_channel/access_entry.py
"""The managed channel's transport, spoken through the plugin's access entry.

`plugins/harness/runtime/access-entry.mjs` is the plugin's one
production entry for connecting to a Harness: it launches the named adapter,
relays every ACP line byte-transparently, and releases what it started only
with an OS-confirmed receipt.  This class is the Server-side counterpart and
owns exactly the protocol needed to use it and nothing else:

  * a channel is not established until the entry's own `connect` has been
    confirmed - no ACP frame is ever written into a connection that was not
    established, and `connect` itself transmits no ACP frame, so the Server
    still never speaks `initialize`/`session/new`/`prompt` for the client;
  * control-plane messages - the entry's replies to our `connect`/`close`
    ops and its lifecycle events - are consumed here and never relayed.  An
    entry failure that carries the id of a client request must not reach the
    client dressed up as an Agent answer, and the receipt of our own close
    is not ACP;
  * the Agent's own end arrives as the entry's `transport_end` event, which
    fires even while the entry process is still alive, and the release path
    goes through the entry's `close` op so the Harness process tree is
    reclaimed by the component that owns it.

Like the direct NDJSON transport it wraps, this class never parses, tracks or
answers an ACP frame: relayed lines pass through byte-for-byte.
"""
from __future__ import annotations

import json
import threading
import uuid
from typing import Any, Callable, Mapping

from ordessa_server.acp_channel.transport import NDJSONTransport, TransportClosed


class _PendingControl:
    """One outstanding control request awaiting the entry's own reply."""

    def __init__(self) -> None:
        self.done = threading.Event()
        self.reply: dict[str, Any] | None = None

    def resolve(self, reply: dict[str, Any] | None) -> None:
        self.reply = reply
        self.done.set()


class AccessEntryTransport:
    """A started plugin entry carrying one ACP channel behind the registry.

    `entry` is the access-entry module path, `node` the interpreter, and
    `adapter` the declared launch reference (`{"command", "args"}`) the entry
    spawns once `connect` names it.  `on_line` receives only the Agent's own
    frames; `on_exit` fires exactly once with a real ending - the entry
    process's return code when the entry itself dies, or the entry's
    `transport_end` reason when the Harness does - and never fabricates a
    frame for requests still in flight.
    """

    def __init__(self, *, node: str, entry: str, harness_id: str, cwd: str,
                 adapter: Mapping[str, Any], on_line: Callable[[str], None],
                 on_exit: Callable[[Any], None],
                 environment: dict[str, str] | None = None,
                 connect_timeout: float = 60.0, close_timeout: float = 12.0) -> None:
        self._harness_id = harness_id
        self._cwd = cwd
        self._adapter = {"command": str(adapter["command"]),
                         "args": [str(arg) for arg in adapter.get("args", [])]}
        self._on_line = on_line
        self._on_exit = on_exit
        self._connect_timeout = connect_timeout
        self._close_timeout = close_timeout
        self._inner = NDJSONTransport(
            argv=[node, entry, "--native"], cwd=cwd, environment=environment,
            on_line=self._classify, on_exit=self._entry_exit,
        )
        self._control_lock = threading.Lock()
        self._control: dict[str, _PendingControl] = {}
        self._end_lock = threading.Lock()
        self._end_reported = False
        self._terminated = False
        self._established = False
        self._fatal_refusal: dict[str, Any] | None = None
        self._events_seen: list[str] = []
        #: The entry's own release report from the last `close` (OS-confirmed or not).
        self.release_report: Any = None
        self._release_lock = threading.Lock()
        self._pending_close: _PendingControl | None = None
        #: Only a confirmed release is cached; a failure stays re-askable.
        self._release_confirmed: dict[str, Any] | None = None

    # -- lifecycle ----------------------------------------------------------

    def start(self) -> "AccessEntryTransport":
        """Spawn the entry and establish the connection before anything rides it.

        A refused or unanswered `connect` reclaims the entry process (which
        reclaims any half-spawned Harness with it) and raises: the registry's
        zero-launch order is kept at this layer too.
        """
        self._inner.start()
        pending = self._request_control({
            "op": "connect", "harness": self._harness_id,
            "launch": dict(self._adapter), "directory": self._cwd,
        })
        if not pending.done.wait(self._connect_timeout):
            self.terminate()
            raise TransportClosed(
                "the access entry never answered its connect request")
        reply = pending.reply
        if reply is None or reply.get("ok") is not True:
            code = ((reply or self._fatal_refusal or {}).get("error") or {}).get(
                "code", "ACP_CONNECT_NO_REPLY")
            self.terminate()
            raise RuntimeError(f"ACP_CHANNEL_CONNECT_REFUSED-{code}")
        self._established = True
        return self

    @property
    def pid(self) -> int | None:
        """The entry process; the Harness it owns is released through `close`."""
        return self._inner.pid

    @property
    def established(self) -> bool:
        return self._established

    def send_line(self, text: str) -> None:
        """One client ACP frame, forwarded as written - the entry relays it."""
        self._inner.send_line(text)

    def terminate(self) -> None:
        """Force the ending: ask the entry's release once, then stop the entry
        whatever it answered.

        This is the crash/server-stop cleanup path.  The client-visible
        release answer is `request_release`, which never trades a missing (or
        negative) receipt for a success claim - here the channel is already
        over on the Agent's side, and leaving the entry running behind it is
        not an option either.
        """
        with self._end_lock:
            self._end_reported = True  # this is our own ending, not a crash report
        with self._release_lock:
            if self._established and self._release_confirmed is None:
                if self._pending_close is None:
                    self._pending_close = self._request_control({"op": "close"})
                pending = self._pending_close
                if pending is not None and pending.done.wait(self._close_timeout):
                    self._pending_close = None
                    self._record_receipt(pending.reply)
        self._stop_entry()

    def request_release(self) -> dict[str, Any]:
        """Ask the entry to release what it started, and answer honestly.

        Confirmation is the entry's own OS-confirmed receipt
        (`ok:true` carrying `released:true`): a refusal, a receipt that
        admits a survivor, or silence is returned as `confirmed: False` -
        never as success, because the component that owns the Harness process
        tree has not proved it is gone.  Only a confirmed release is kept: a
        failed receipt leaves the next explicit retry free to ask the entry
        again, which is the whole point of a retryable release.  A close that
        was never answered keeps its pending request - the retry waits on the
        same one, so a late receipt can still confirm and no second close is
        written behind the first.  The entry - the only manager of the
        Harness it owns - is never destroyed for answering badly or being
        slow.
        """
        with self._release_lock:
            if self._release_confirmed is not None:
                return self._release_confirmed
            if not self._established:
                return {"confirmed": False, "reason": "never-established",
                        "report": None}
            pending = self._pending_close
            if pending is None:
                with self._end_lock:
                    self._end_reported = True  # from here this is our own ending
                pending = self._request_control({"op": "close"})
                if pending is None:
                    return {"confirmed": False, "reason": "entry-unreachable",
                            "report": None}
                self._pending_close = pending
            if not pending.done.wait(self._close_timeout):
                # Not cached and the pending is kept: the receipt may still be
                # on its way, and a retry waits on the same request.
                return {"confirmed": False, "reason": "close-timeout",
                        "report": None}
            self._pending_close = None
            outcome = self._record_receipt(pending.reply)
            if outcome["confirmed"]:
                self._stop_entry()
            return outcome

    def _record_receipt(self, reply: dict[str, Any] | None) -> dict[str, Any]:
        """Turn one `close` reply into the honest release outcome.

        Only the confirmation is cached; every failure shape is recomputed
        per attempt, so a later retry can genuinely ask again.
        """
        if self._release_confirmed is not None:
            return self._release_confirmed
        if reply is None:
            return {"confirmed": False, "reason": "entry-exited-without-receipt",
                    "report": None}
        report = reply.get("result") if isinstance(reply.get("result"), dict) \
            else (reply.get("error") or reply)
        self.release_report = report
        if reply.get("ok") is True and report.get("released") is True:
            self._release_confirmed = {
                "confirmed": True, "reason": "released", "report": report}
            return self._release_confirmed
        if reply.get("ok") is not True:
            code = str(((reply.get("error") or {}).get("code")) or "unknown")
            return {"confirmed": False, "reason": f"close-refused-{code}",
                    "report": report}
        return {"confirmed": False, "reason": "release-unconfirmed",
                "report": report}

    def _stop_entry(self) -> None:
        """Stop the entry process exactly once (stdin EOF, then escalation)."""
        with self._end_lock:
            if self._terminated:
                return
            self._terminated = True
        self._inner.terminate()

    # -- the entry's stdout, classified ---------------------------------------

    def _classify(self, line: str) -> None:
        text = line.strip()
        if text.startswith("{"):
            try:
                message = json.loads(text)
            except ValueError:
                message = None
            if isinstance(message, dict) and "jsonrpc" not in message:
                if message.get("ok") is True or message.get("ok") is False:
                    self._resolve_control(message)
                    return
                if isinstance(message.get("event"), str):
                    self._handle_event(message)
                    return
        # Anything that declares itself part of the JSON-RPC conversation -
        # and every line the entry cannot be speaking as itself - is the
        # Agent's frame, relayed exactly as it arrived.
        self._on_line(line)

    def _request_control(self, request: dict[str, Any]) -> _PendingControl | None:
        control_id = uuid.uuid4().hex
        pending = _PendingControl()
        with self._control_lock:
            self._control[control_id] = pending
        try:
            self._inner.send_line(json.dumps({"id": control_id, **request}))
        except TransportClosed:
            with self._control_lock:
                self._control.pop(control_id, None)
            return None
        return pending

    def _resolve_control(self, message: dict[str, Any]) -> None:
        control_id = message.get("id")
        pending = None
        with self._control_lock:
            if isinstance(control_id, str):
                pending = self._control.pop(control_id, None)
        if pending is None:
            # A reply to a request we never made (or an entry-wide failure):
            # control-plane, never an Agent answer - kept only as evidence.
            if message.get("ok") is False and self._fatal_refusal is None:
                self._fatal_refusal = message
            return
        pending.resolve(message)

    def _abandon_control(self) -> None:
        with self._control_lock:
            outstanding, self._control = self._control, {}
        for pending in outstanding.values():
            pending.resolve(None)

    def _handle_event(self, message: dict[str, Any]) -> None:
        event = str(message.get("event"))
        self._events_seen.append(event)
        if event == "transport_end":
            # The entry stays alive after its Harness dies, so a `close` asked
            # of it is still answerable: control waiters are abandoned only
            # when the entry process itself is gone (`_entry_exit`).
            reason = str((message.get("data") or {}).get("reason")
                         or "transport-end")
            self._report_end(reason)
        # `process_release_unconfirmed` is the entry saying it could not prove
        # the OS stopped a process it owns: kept as evidence, never relayed,
        # and never rewritten into a claim that the release succeeded.
        # `stderr`/`transport_malformed` are the Agent's diagnostics on the
        # control plane - the ACP relay stays frame-only.

    def _entry_exit(self, code: int | None) -> None:
        # The inner reader already proved the entry process gone; anything it
        # was asked to release is now unrecoverable, so no waiter stays parked.
        self._abandon_control()
        self._report_end(code)

    def _report_end(self, code: Any) -> None:
        with self._end_lock:
            if self._end_reported:
                return
            self._end_reported = True
        try:
            self._on_exit(code)
        except Exception:  # noqa: BLE001 - the ledger owner must not be re-entered
            pass
        # The entry may still be running behind a dead Harness (that is the
        # normal crash shape); its release is this transport's to ask for, off
        # the reader thread so the entry's own answer can arrive.
        threading.Thread(target=self.terminate, daemon=True).start()
